p2p/src/network/scheduler/p2p_network_scheduler_reducer.rs

```diff
@@ -64,148 +132,480 @@ impl P2pNetworkSchedulerState {
+                let p2p_state: &P2pState = state.substate()?;
+
+                match result {
+                    Ok(()) => {
+                        dispatcher.push(P2pNetworkSchedulerEffectfulAction::OutgoingDidConnect {
+                            addr: *addr,
+                        });
+                    }
+                    Err(error) => {
+                        let Some((peer_id, peer_state)) = p2p_state.peer_with_connection(*addr)
+                        else {
+                            bug_condition!(
+                                "outgoing connection to {addr} failed, but there is no peer for it"
+                            );
+                            return Ok(());
+                        };
+                        if matches!(
+                            peer_state.status,
+                            P2pPeerStatus::Connecting(P2pConnectionState::Outgoing(_))
+                        ) {
+                            dispatcher.push(P2pConnectionOutgoingAction::FinalizeError {
+                                peer_id,
+                                error: error.to_string(),
+                            });
+                        }
+                    }
+                }
+                Ok(())
             }
-            P2pNetworkSchedulerAction::IncomingDataIsReady { .. } => {}
             P2pNetworkSchedulerAction::IncomingDataDidReceive { result, addr } => {
-                let Some(state) = self.connections.get_mut(addr) else {
-                    return;
+                if let Some(state) = scheduler_state.connections.get_mut(addr) {
+                    if let Ok(data) = result {
+                        state.consume(data.len());
+                    }
                 };
 
-                if let Ok(data) = result {
-                    state.consume(data.len());
+                let dispatcher = state_context.into_dispatcher();
+                match result {
+                    Ok(data) => {
+                        dispatcher.push(P2pNetworkPnetAction::IncomingData {
+                            addr: *addr,
+                            data: data.clone(),
+                        });
+                    }
+                    Err(error) => dispatcher.push(P2pNetworkSchedulerAction::Error {
+                        addr: *addr,
+                        error: P2pNetworkConnectionError::MioError(error.clone()),
+                    }),
                 }
+                Ok(())
             }
             P2pNetworkSchedulerAction::SelectDone {
                 addr,
                 kind,
                 protocol,
                 incoming,
                 expected_peer_id,
-                ..
             } => {
-                let Some(connection) = self.connections.get_mut(addr) else {
-                    return;
+                if let Some(connection) = scheduler_state.connections.get_mut(addr) {
```

Comment:
General observation: now that "effects" and reducer code are together, what happens when we can't find connection state for some address? Does still make sense to execute the "effects" part and dispatch further actions? Is it a bug condition not being able to find associated state to an address?

Reply:
In most cases it doesn't make sense to execute "effects" if reducer is unable to be executed, but there are cases in which we need to execute "effects" even if reducer fails, most of these cases are to dispatch error, so just in case I try to port the reducer so even if it can't execute to still go to effects

Reply:
such cases where the reducer cannot be executed and the dispatches must not happen either should be filtered by enabling conditions right?

Reply:
btw this whole block of code here is another good candidate for being moved to a helper

Reply:
I have moved both parts to helpers, it looks much better and cleaner now